### PR TITLE
Fixed Bgen 1.1 bugs, made sampleqc faster

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 1]
```diff
@@ -40,6 +37,7 @@ class BgenRecordV11(compressed: Boolean,
     new Iterable[Genotype] {
       def iterator = new Iterator[Genotype] {
         var i = 0
+        byteReader.seek(0)
```
Comment (tpoterba) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: this is the bugfix

File: src/main/scala/is/hail/methods/VariantQC.scala [hunk 2]
```diff
@@ -113,57 +107,6 @@ class VariantQCCombiner extends Serializable {
     (divOption(LH.getNumericalMean, n), LH.exactMidP(nAB))
   }
 
-  def emit(sb: mutable.StringBuilder) {
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/VariantQC.scala]: unused

File: src/main/scala/is/hail/methods/SampleQC.scala [hunk 3]
```diff
@@ -140,68 +137,6 @@ class SampleQCCombiner(val keepStar: Boolean) extends Serializable {
     this
   }
 
-  def emitSC(sb: mutable.StringBuilder, sc: StatCounter) {
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/SampleQC.scala]: stale + unused

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 4]
```diff
@@ -61,17 +59,10 @@ class BgenRecordV11(compressed: Boolean,
...
             } else {
-              setWarning(gpSumGreaterThanTolerance)
               noCall
             }
-          } else {
-            if (dsum == 0)
-              setWarning(gpNoCall)
-            else
-              setWarning(gpSumLessThanTolerance)
-
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Why did you remove the warnings for v1.1?

**Comment by johnc1231** [comment]
Addresses #1954

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 4 above)
Comment (tpoterba) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: they weren't used. Anywhere

Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: They were used. BgenLoader line 44.

File: src/main/scala/is/hail/methods/SampleQC.scala [hunk 5]
```diff
@@ -235,16 +177,21 @@ object SampleQC {
...
-                g.gt
-                  .filter(_ > 0)
-                  .foreach(call => Genotype.gtPair(call).alleleIndices.filter(_ > 0).foreach(x => acc(x - 1) += 1)
-                  )
-                acc
-            })
-            for ((g, i) <- gs.iterator.zipWithIndex)
-              acc(i).merge(v, ACs, g)
+            val acs = Array.fill(v.nAlleles)(0)
+            gs.foreach { g =>
```
Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: You could replace `foreach` with a while loop to make it faster.

File: src/main/scala/is/hail/methods/SampleQC.scala [hunk 6]
```diff
@@ -235,16 +177,21 @@ object SampleQC {
...
+              if (g.unboxedGT >= 0) {
+                val gtPair = Genotype.gtPair(g.unboxedGT)
+                acs(gtPair.j) += 1
+                acs(gtPair.k) += 1
+              }
+            }
+
+            var i = 0
+            gs.foreach { g =>
```
Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: same here.

File: src/main/scala/is/hail/methods/SampleQC.scala [hunk 7]
```diff
@@ -69,55 +69,59 @@ class SampleQCCombiner(val keepStar: Boolean) extends Serializable {
...
+        mergeAllele(gtPair.j)
+        mergeAllele(gtPair.k)
 
-    if (g.isCalled) {
-      g.dp.foreach { v =>
-        dpSC.merge(v)
-      }
-      g.gq.foreach { v =>
-        gqSC.merge(v)
+        if (gtPair.j == gtPair.k)
```
Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: This should be `gtPair.j != gtPair.k`.

File: src/main/scala/is/hail/methods/SampleQC.scala [hunk 8]
```diff
@@ -69,55 +69,59 @@ class SampleQCCombiner(val keepStar: Boolean) extends Serializable {
...
+          if (ai > 0 && (keepStar || !v.altAlleles(ai - 1).isStar)) {
 
-        if (!nonRefAlleleIndices.isEmpty) {
-          nonRefAlleleIndices.foreach({
-            ai =>
-              val altAllele = v.altAlleles(ai - 1)
-              if (altAllele.isSNP) {
+            val altAllele = v.altAlleles(ai - 1)
+            altAllele.altAlleleType match {
+              case AltAlleleType.SNP =>
```
Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: Should this be if/else instead of match for performance considerations?

File: src/main/scala/is/hail/methods/SampleQC.scala (hunk 7 above)
Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: In the original, nHet and nHomVar are not incremented if keepStar is false. Make consistent with original.

Comment (jigold) [inline src/main/scala/is/hail/methods/SampleQC.scala]: We should have handwritten tests for all parameters computed by sampleQC and variantQC for a small example. This should have been caught by a test. I see some sample QC output is tested in `FilterSuite`, but only `nCalled` and `dpMean`.

File: src/main/scala/is/hail/methods/SampleQC.scala (hunk 5 above)
Comment (tpoterba) [inline src/main/scala/is/hail/methods/SampleQC.scala]: while won't speed up iterable.foreach, I think

File: src/main/scala/is/hail/methods/SampleQC.scala (hunk 7 above)
Comment (tpoterba) [inline src/main/scala/is/hail/methods/SampleQC.scala]: good catch!

**Comment by jigold** [comment]
Also, don't forget about the accumulators in GenLoader.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 9]
```diff
@@ -217,4 +217,13 @@ class LoadBgenSuite extends SparkSuite {
     hc.indexBgen(bgen)
     hc.importBgen(bgen, Option(sample)).count()
   }
+
+  @Test def testReIterate() {
+    val vds = hc.importBgen("src/test/resources/example.v11.bgen", Some("src/test/resources/example.sample"))
```
Comment (jigold) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: You need to index the bgen file first. Or we can add the index to the repository.
